presto-main/src/main/java/io/prestosql/sql/analyzer/Scope.java

```diff
@@ -99,6 +103,95 @@ public RelationType getRelationType()
         return relation;
     }
 
+    public Optional<AsteriskedIdentifierChainBasis> resolveAsteriskedIdentifierChainBasis(QualifiedName identifierChain)
+    {
+        int length = identifierChain.getParts().size();
+        if (length == 1) {
+            return resolveTableBasis(identifierChain);
+        }
+        if (length <= 3) {
+            return resolveTableOrFieldBasis(identifierChain);
+        }
+        return resolveFieldBasis(identifierChain);
+    }
+
+    private Optional<AsteriskedIdentifierChainBasis> resolveTableBasis(QualifiedName identifierChain)
+    {
+        Optional<AsteriskedIdentifierChainBasis> basis = resolveTableBasisLocally(identifierChain);
+        if (!basis.isPresent() && getOuterQueryParent().isPresent()) {
+            basis = getOuterQueryParent().get().resolveTableBasisLocally(identifierChain);
+        }
+        return basis;
+    }
+
+    private Optional<AsteriskedIdentifierChainBasis> resolveFieldBasis(QualifiedName identifierChain)
+    {
+        Optional<AsteriskedIdentifierChainBasis> basis = resolveFieldBasisLocally(identifierChain);
+        if (!basis.isPresent() && getOuterQueryParent().isPresent()) {
+            basis = getOuterQueryParent().get().resolveFieldBasisLocally(identifierChain);
+        }
+        return basis;
+    }
+
+    private Optional<AsteriskedIdentifierChainBasis> resolveTableOrFieldBasis(QualifiedName identifierChain)
+    {
+        Optional<AsteriskedIdentifierChainBasis> tableBasis = resolveTableBasisLocally(identifierChain);
+        Optional<AsteriskedIdentifierChainBasis> fieldBasis = resolveFieldBasisLocally(identifierChain);
+        if (tableBasis.isPresent() && fieldBasis.isPresent()) {
+            return Optional.of(AMBIGUOUS);
+        }
+        if (tableBasis.isPresent()) {
+            return Optional.of(TABLE);
+        }
+        if (fieldBasis.isPresent()) {
+            return Optional.of(FIELD);
+        }
+        if (getOuterQueryParent().isPresent()) {
+            return getOuterQueryParent().get().resolveTableOrFieldBasis(identifierChain);
+        }
+        return Optional.empty();
+    }
+
+    private Optional<AsteriskedIdentifierChainBasis> resolveTableBasisLocally(QualifiedName identifierChain)
+    {
+        if (relation.getAllFields().stream()
+                .anyMatch(field -> field.getRelationAlias().isPresent() && field.getRelationAlias().get().equals(identifierChain))) {
```

Comment:
Actually, the spec says this about it:

<img width="944" alt="Screen Shot 2019-10-16 at 10 40 01 PM" src="https://user-images.githubusercontent.com/14387/66980483-e8d2c100-f065-11e9-8ca2-735f79e852bc.png">

So that query should fail because there can't be two `t` range variables in the `FROM` clause.

Reply:
Here's another fun one that should be disallowed:

```
SELECT orders.* 
FROM orders, lineitem AS orders
```

Reply:
So such ambiguity should be sorted out when the `FROM` clause is analysed (that is before we start to analyse `SELECT` items).

Reply:
> To fix this, we'd need to record the available table and correlation names in the scope object. It's a much bigger change and probably out of scope. for this PR.

If we followed what the spec says about the `FROM` clause, there would be no ambiguous names in scope. However, I don't think we want to do it - it would break queries that are intuitively correct. And still the `WITH` clause might introduce ambiguity.

And if we want to deal with ambiguity of aliases/table names in scope, or even notice it, then we need to restructure `Scope` to track fields' origin instead of having a single `RelationType`.

Reply:
Even if we want to do it lazily, we need to record the individual aliases and table names. Otherwise, we can’t tell if there’s more than one at the point of resolution. Also, thinking about it more, RelationType is probably the wrong abstraction for how we need to do resolution anyway.